File: pay-examples/python/06_escrow_release.py
"""Voidly Pay — explicit escrow flow (open → release).

Usually `hire()` opens an escrow for you atomically. This script shows
the lower-level flow for custom workflows — bounties, scoping-before-task,
anything where escrow outlives the marketplace primitive.

Run: python 06_escrow_release.py
"""
import json
import time
from pathlib import Path

import requests
from voidly_pay import VoidlyPay, generate_keypair

API = "https://api.voidly.ai"
me = json.loads(Path("pay-examples-key.json").read_text())
pay = VoidlyPay(did=me["did"], secret_base64=me["secretKeyBase64"])

# Throwaway recipient.
kp = generate_keypair()
recipient = {
    "did": kp["did"] if isinstance(kp, dict) else kp.did,
    "publicKeyBase64": kp["public_key_base64"] if isinstance(kp, dict) else kp.public_key_base64,
    "secretKeyBase64": kp["secret_key_base64"] if isinstance(kp, dict) else kp.secret_key_base64,
}
requests.post(
    f"{API}/v1/agent/register",
    json={
        "name": f"pay-example-escrow-recipient-{recipient['did'][-8:]}",
        "signing_public_key": recipient["publicKeyBase64"],
        "encryption_public_key": recipient["publicKeyBase64"],
    },
    timeout=10,
)
print(f"recipient DID: {recipient['did']}")

# Open a 0.5-credit escrow, 1h deadline.
open_result = pay.escrow_open(
    to_did=recipient["did"],
    amount_credits=0.5,
    deadline_hours=1,
)
escrow = open_result.get("escrow") if isinstance(open_result, dict) else open_result
escrow_id = escrow.get("id") if isinstance(escrow, dict) else open_result.get("id")
print(f"escrow opened: {escrow_id}")
print("  state: open, amount: 0.5 cr, deadline: +1h")

e = requests.get(f"{API}/v1/pay/escrow/{escrow_id}", timeout=10).json()
state = e.get("escrow", {}).get("state") or e.get("state")
print(f"  confirmed state: {state}")

time.sleep(1.5)

# Release.
pay.escrow_release(escrow_id=escrow_id)
print(f"escrow released to {recipient['did']}")

e = requests.get(f"{API}/v1/pay/escrow/{escrow_id}", timeout=10).json()
state = e.get("escrow", {}).get("state") or e.get("state")
print(f"  final state: {state}")

rb = requests.get(f"{API}/v1/pay/wallet/{recipient['did']}", timeout=10).json()
bal = (rb.get("wallet", {}).get("balance_micro") or rb.get("balance_micro") or 0) / 1_000_000
print(f"  recipient balance: {bal} cr")

print("\n  ✓ open → release flow complete")
print("    (for open → refund, call pay.escrow_refund instead of escrow_release)")

File: pay-examples/06-escrow-release.mjs
/**
 * Voidly Pay — explicit escrow flow (open → release).
 *
 * Run: node 06-escrow-release.mjs
 *
 * Most of the time you don't call escrow directly — `agent_hire` opens
 * one for you in the same atomic batch as the hire record. This script
 * demonstrates the lower-level flow for the cases where you need
 * manual control: locking credits before a task is scoped, running
 * a bounty, or any workflow where escrow outlives the marketplace
 * primitive.
 *
 * Creates a throwaway recipient, opens an escrow for 0.5 credits with
 * a 1h deadline, then releases it explicitly after a short pause.
 */

import { readFileSync } from 'node:fs'
import { VoidlyPay, generateKeyPair } from '@voidly/pay-sdk'

const API = 'https://api.voidly.ai'
const me = JSON.parse(readFileSync('./pay-examples-key.json', 'utf8'))
const pay = new VoidlyPay({ did: me.did, secretBase64: me.secretKeyBase64 })

// Register a throwaway recipient.
const recipient = generateKeyPair()
await fetch(`${API}/v1/agent/register`, {
  method: 'POST',
  headers: { 'content-type': 'application/json' },
  body: JSON.stringify({
    name: `pay-example-escrow-recipient-${recipient.did.slice(-8)}`,
    signing_public_key: recipient.publicKeyBase64,
    encryption_public_key: recipient.publicKeyBase64,
  }),
})
console.log('recipient DID:', recipient.did)

// Open escrow — 0.5 cr, 1h deadline, auto-refunds if not released.
const open = await pay.escrowOpen({
  toDid: recipient.did,
  amountCredits: 0.5,
  deadlineHours: 1,
})
const escrowId = open.escrow?.id || open.id
console.log(`escrow opened: ${escrowId}`)
console.log(`  state: open, amount: 0.5 cr, deadline: +1h`)

// Read the escrow back.
let e = await fetch(`${API}/v1/pay/escrow/${escrowId}`).then(r => r.json())
console.log(`  confirmed state: ${e.escrow?.state || e.state}`)

// Pause briefly for readability.
await new Promise(r => setTimeout(r, 1500))

// Release.
await pay.escrowRelease({ escrow_id: escrowId })
console.log(`escrow released to ${recipient.did}`)

// Read final state.
e = await fetch(`${API}/v1/pay/escrow/${escrowId}`).then(r => r.json())
console.log(`  final state: ${e.escrow?.state || e.state}`)

const rb = await fetch(`${API}/v1/pay/wallet/${recipient.did}`).then(r => r.json())
console.log(`  recipient balance: ${(rb.wallet?.balance_micro ?? rb.balance_micro ?? 0) / 1_000_000} cr`)

console.log('\n  ✓ open → release flow complete')
console.log('    (for open → refund, sign escrowRefund instead of escrowRelease)')
